File: app/src/main/java/com/github/adrijanrogan/etiketa/ui/browse/BrowseAdapter.java
package com.github.adrijanrogan.etiketa.ui.browse;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.github.adrijanrogan.etiketa.R;

import java.io.File;

class BrowseAdapter extends RecyclerView.Adapter<BrowseAdapter.FileViewHolder> {

    private File[] files;
    private AdapterCallback callback;


    class FileViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        ImageView fileIcon;
        TextView fileName;
        TextView subFiles;

        FileViewHolder(View root) {
            super(root);
            root.setOnClickListener(this);
            this.fileIcon = root.findViewById(R.id.holder_icon);
            this.fileName = root.findViewById(R.id.holder_file_name);
            this.subFiles = root.findViewById(R.id.holder_sub_files);
        }

        @Override
        public void onClick(View v) {
            callback.onClickFile(getLayoutPosition());
        }
    }



    public BrowseAdapter(File[] files, AdapterCallback callback) {
        this.files = files;
        this.callback = callback;
    }

    @NonNull
    @Override
    public FileViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View root = LayoutInflater.
                from(parent.getContext()).inflate(R.layout.holder_file, parent, false);
        return new FileViewHolder(root);
    }

    // Ta metoda se pokliče za vsak element v seznamu, ko bo element postal viden.
    // Tu določimo ikono, ki jo vidi uporabnik, glede na to, ali je to mapa, datoteka
    // ali glasbena datoteka. Uporabnik vidi tudi ime datoteke.
    @Override
    public void onBindViewHolder(@NonNull FileViewHolder holder, int position) {
        File file = files[position];
        String fileName = file.getName();

        holder.fileName.setText(fileName);
        // Pravilno sklanjamo glede na stevilo datotek v mapi.
        if (file.isDirectory()) {
            holder.subFiles.setVisibility(View.VISIBLE);
            switch (file.listFiles().length) {
                case 1:
                    holder.subFiles.setText("1 datoteka");
                    break;
                case 2:
                    holder.subFiles.setText("2 datoteki");
                    break;
                case 3:
                case 4:
                    holder.subFiles.setText(file.listFiles().length + " datoteke");
                    break;
                default:
                    holder.subFiles.setText(file.listFiles().length + " datotek");
                    break;
            }
        } else {
            holder.subFiles.setVisibility(View.GONE);
        }


        if (file.isDirectory()) {
            holder.fileIcon.setImageResource(R.drawable.ic_folder_black_24dp);
        } else if (fileName.endsWith("mp3") || fileName.endsWith("flac")) {
            holder.fileIcon.setImageResource(R.drawable.ic_music_note_black_24dp);
        } else {
            holder.fileIcon.setImageResource(R.drawable.ic_file_black_24dp);
        }

    }

    @Override
    public int getItemCount() {
        return files.length;
    }

}

File: app/src/main/java/com/github/adrijanrogan/etiketa/ui/browse/BrowseAdapter.kt
package com.github.adrijanrogan.etiketa.ui.browse

import android.content.Context
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.github.adrijanrogan.etiketa.R
import java.io.File

class BrowseAdapter(private val context: Context, private val callback: BrowserCallback) :
        ListAdapter<File, BrowseAdapter.FileViewHolder>(DIFF_CALLBACK) {


    inner class FileViewHolder(val root: View) : RecyclerView.ViewHolder(root) {
        var fileIcon: ImageView = root.findViewById(R.id.holder_icon)
        var fileName: TextView = root.findViewById(R.id.holder_file_name)
        var subFiles: TextView = root.findViewById(R.id.holder_sub_files)
        var options: ImageView = root.findViewById(R.id.holder_options)
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FileViewHolder {
        val root = LayoutInflater.from(parent.context).inflate(R.layout.holder_file, parent, false)
        return FileViewHolder(root)
    }


    override fun onBindViewHolder(holder: FileViewHolder, position: Int) {
        val file = getItem(position)
        val fileName = file.name

        holder.root.setOnClickListener { callback.onClickFile(getItem(position)) }
        holder.options.setOnClickListener { showPopupMenu(it, getItem(position)) }

        if (file.isHidden) {
            holder.fileName.alpha = 0.7F
            holder.subFiles.alpha = 0.7F
            holder.fileIcon.imageAlpha = 127
        } else {
            holder.fileName.alpha = 1.0F
            holder.subFiles.alpha = 1.0F
            holder.fileIcon.imageAlpha = 255
        }

        holder.fileName.text = fileName
        if (file.isDirectory) {
            holder.subFiles.visibility = View.VISIBLE
            val count = file.listFiles().size
            if (count == 0) holder.subFiles.text = context.getString(R.string.numberOfFilesEmpty)
            else holder.subFiles.text = context.resources.getQuantityString(
                    R.plurals.numberOfFiles, count, count)
        } else {
            holder.subFiles.visibility = View.GONE
        }

        // TODO: Move this to something like an utility class


        if (file.isDirectory) {
            holder.fileIcon.setImageResource(R.drawable.ic_outline_folder)
        } else if (fileName.endsWith("mp3") || fileName.endsWith("flac")) {
            holder.fileIcon.setImageResource(R.drawable.ic_outline_music_note)
        } else if (fileName.endsWith("png") || fileName.endsWith("jpg") ||
                fileName.endsWith("jpeg")) {
            holder.fileIcon.setImageResource(R.drawable.ic_outline_photo)
        } else {
            holder.fileIcon.setImageResource(R.drawable.ic_outline_file)
        }

    }

    private fun showPopupMenu(view: View, file: File) {
        val popup = PopupMenu(context, view)
        popup.menuInflater.inflate(R.menu.browser_file_menu, popup.menu)
        popup.setOnMenuItemClickListener { onPopupMenuItemClicked(it, file) }
        popup.show()
    }


    private fun onPopupMenuItemClicked(menuItem: MenuItem, file: File): Boolean {
        when (menuItem.itemId) {
            R.id.browser_menu_show_info -> {
                callback.showFileInfoDialog(file)
            }
            R.id.browser_menu_as_album ->
                Toast.makeText(context, "TODO: Treat as album", Toast.LENGTH_LONG).show()
        }
        return true
    }


    companion object {
        val DIFF_CALLBACK: DiffUtil.ItemCallback<File> = object:DiffUtil.ItemCallback<File>() {
            // Must compare paths, otherwise folders with same names cause trouble
            override fun areItemsTheSame(oldItem: File, newItem: File): Boolean {
                return oldItem.name == newItem.name &&
                        oldItem.absolutePath == newItem.absolutePath
            }

            override fun areContentsTheSame(oldItem: File, newItem: File): Boolean {
                return areItemsTheSame(oldItem, newItem)
            }
        }
    }
}